src/gc/gc.h

```diff
@@ -158,105 +133,21 @@ namespace SVR {
-    friend class SVR::gc_heap;
-    friend class SVR::GCHeap;
-#endif // defined(FEATURE_SVR_GC)
-    friend struct ClassDumpInfo;
-
-    uint8_t*       alloc_ptr;
-    uint8_t*       alloc_limit;
-    int64_t        alloc_bytes; //Number of bytes allocated on SOH by this context
-    int64_t        alloc_bytes_loh; //Number of bytes allocated on LOH by this context
-#if defined(FEATURE_SVR_GC)
-    SVR::GCHeap*   alloc_heap;
-    SVR::GCHeap*   home_heap;
-#endif // defined(FEATURE_SVR_GC)
-    int            alloc_count;
-public:
-
-    void init()
-    {
-        LIMITED_METHOD_CONTRACT;
-
-        alloc_ptr = 0;
-        alloc_limit = 0;
-        alloc_bytes = 0;
-        alloc_bytes_loh = 0;
-#if defined(FEATURE_SVR_GC)
-        alloc_heap = 0;
-        home_heap = 0;
-#endif // defined(FEATURE_SVR_GC)
-        alloc_count = 0;
-    }
-};
-
-struct ScanContext
-{
-    Thread* thread_under_crawl;
-    int thread_number;
-    uintptr_t stack_limit; // Lowest point on the thread stack that the scanning logic is permitted to read
-    BOOL promotion; //TRUE: Promotion, FALSE: Relocation.
-    BOOL concurrent; //TRUE: concurrent scanning 
-#if CHECK_APP_DOMAIN_LEAKS || defined (FEATURE_APPDOMAIN_RESOURCE_MONITORING) || defined (DACCESS_COMPILE)
-    AppDomain *pCurrentDomain;
-#endif //CHECK_APP_DOMAIN_LEAKS || FEATURE_APPDOMAIN_RESOURCE_MONITORING || DACCESS_COMPILE
-
-#ifndef FEATURE_REDHAWK
-#if defined(GC_PROFILING) || defined (DACCESS_COMPILE)
-    MethodDesc *pMD;
-#endif //GC_PROFILING || DACCESS_COMPILE
-#endif // FEATURE_REDHAWK
-#if defined(GC_PROFILING) || defined(FEATURE_EVENT_TRACE)
-    EtwGCRootKind dwEtwRootKind;
-#endif // GC_PROFILING || FEATURE_EVENT_TRACE
-    
-    ScanContext()
-    {
-        LIMITED_METHOD_CONTRACT;
-
-        thread_under_crawl = 0;
-        thread_number = -1;
-        stack_limit = 0;
-        promotion = FALSE;
-        concurrent = FALSE;
-#ifdef GC_PROFILING
-        pMD = NULL;
-#endif //GC_PROFILING
-#if defined(GC_PROFILING) || defined(FEATURE_EVENT_TRACE)
-        dwEtwRootKind = kEtwGCRootKindOther;
-#endif // GC_PROFILING || FEATURE_EVENT_TRACE
-    }
-};
-
-typedef BOOL (* walk_fn)(Object*, void*);
-typedef void (* gen_walk_fn)(void *context, int generation, uint8_t *range_start, uint8_t * range_end, uint8_t *range_reserved);
-
-#if defined(GC_PROFILING) || defined(FEATURE_EVENT_TRACE)
-struct ProfilingScanContext : ScanContext
-{
-    BOOL fProfilerPinned;
-    void * pvEtwContext;
-    void *pHeapId;
-    
-    ProfilingScanContext(BOOL fProfilerPinnedParam) : ScanContext()
-    {
-        LIMITED_METHOD_CONTRACT;
-
-        pHeapId = NULL;
-        fProfilerPinned = fProfilerPinnedParam;
-        pvEtwContext = NULL;
-#ifdef FEATURE_CONSERVATIVE_GC
-        // To not confuse GCScan::GcScanRoots
-        promotion = g_pConfig->GetGCConservative();
+#ifdef FEATURE_SVR_GC
+    inline SVR::GCHeap*& alloc_heap() { return reinterpret_cast<SVR::GCHeap*&>(gc_reserved_1); }
```

Comment:
The calls to alloc_heap where you assign to the returned value look confusing to me. It seems it would be better to split this function into a getter and a setter. 
The same for the home_heap below.